common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -34,6 +34,15 @@
  * Utility class for collation-aware UTF8String operations.
  */
 public class CollationAwareUTF8String {
+
+  public static boolean lowercaseMatchAt(final UTF8String l, final UTF8String r,
+      int pos, int len) {
```

Comment:
I think this API is a bit problematic. We want `len` to be the number of characters of `r` all the time, right?

Reply:
no, `len` is used for `l` and we don't know in advance whether `len` will be equal to `r.numChars()` (it actually won't be in these special cases)

however, this is irrelevant because I need to change the approach overall, so `len` will likely be removed from this API either way